### PR TITLE
Add test folder and update the cmake file

### PR DESCRIPTION
**Pull request by Robchaves**
A new folder is added with a couple of files, these are for test, also the cmake file's updated to include the gtest project.

### CODE REVIEW

File: localMatchApp/CMakeLists.txt [hunk 1]
```diff
@@ -48,4 +48,58 @@ ADD_EXECUTABLE(local_match ${code_SOURCES}
...
+###################################################################
+# Gtest section
+
+# Create a variable with the name of the project and "_test"
+SET(TEST_LOCAL_MATCH ${PROJECT_NAME}_test)
+
+# Define the code language as C++ 11
+ADD_DEFINITIONS(-DGTEST_LANGUAGE_CXX11)
+
+# Search for a package required for execute gtest
```
Comment (luvalverde) [inline localMatchApp/CMakeLists.txt]: ... for executing gtest

File: localMatchApp/headers/Login.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    wrongPassword
+  };
+
+ public:
+  Login ( );
+  ~Login ( );
+  Login ( const Login& ) = default;
+  Login& operator= ( const Login& ) = default;
+
+  connectionMessage signUp ( const std::string i_user, const std::string i_password,
```
Comment (luvalverde) [inline localMatchApp/headers/Login.h]: missing documentation for all these functions and variables of .h file

File: localMatchApp/sources/Login.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
+#include "Login.h"
+
+using namespace std;
+
+Login::Login ( ) {
+  // Connect to database.
```
Comment (luvalverde) [inline localMatchApp/sources/Login.cpp]: Add a TODO comment for this pending task.

File: localMatchApp/sources/Login.cpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+Login::~Login ( ) {
+  // Disconnect to database.
```
Comment (luvalverde) [inline localMatchApp/sources/Login.cpp]: Add a TODO comment for this pending task.

File: localMatchApp/sources/Login.cpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+Login::connectionMessage Login::signUp ( const string i_user, const string i_password,
+                                         const string i_repeatedPassword ) {
+  // Send information to database.
+  string userInDatabase = "";
+
+  if ( hasCorrectFormat ( i_password ) && hasCorrectFormat ( i_repeatedPassword ) ) {
+    if ( i_user.compare ( userInDatabase ) == 0 ) {
+      if ( i_password == i_repeatedPassword ) {
+        // Set user and password in table
```
Comment (luvalverde) [inline localMatchApp/sources/Login.cpp]: Add a TODO comment for this pending task.
Apply this comment is all your PR.

File: localMatchApp/sources/Login.cpp [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+Login::~Login ( ) {
+  // Disconnect to database.
+}
+
+Login::connectionMessage Login::signUp ( const string i_user, const string i_password,
+                                         const string i_repeatedPassword ) {
+  // Send information to database.
+  string userInDatabase = "";
+
+  if ( hasCorrectFormat ( i_password ) && hasCorrectFormat ( i_repeatedPassword ) ) {
```
Comment (luvalverde) [inline localMatchApp/sources/Login.cpp]: To avoid nesting IF statements, somethings it is easier to check the failure condition like:
```c
if ( !hasCorrectFormat ( i_password ) || !hasCorrectFormat ( i_repeatedPassword ) ) {
  return connectionMessage::wrongPassword;
}

if ( i_password != i_repeatedPassword ) {
  return connectionMessage::notEqualPasswords;
}

if ( i_user.compare ( userInDatabase ) != 0 ) {
  return connectionMessage::ocupatedUser;
}

// Set user and password in table
return connectionMessage::goodEntry;
```

File: localMatchApp/sources/Login.cpp [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    } else {
+      return connectionMessage::ocupatedUser;
+    }
+  } else {
+    return connectionMessage::wrongPassword;
+  }
+}
+
+Login::connectionMessage Login::entry ( const string i_user, const string i_password ) {
+  // Search the user and compare password
```
Comment (luvalverde) [inline localMatchApp/sources/Login.cpp]: Add a TODO comment for this pending task.

File: localMatchApp/sources/Login.cpp [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+  // Query asking for password based on user
+  string userInDatabase = "";
+  string passInDatabase = "";
+
```
Comment (luvalverde) [inline localMatchApp/sources/Login.cpp]: Add a comment describing what those IF statements do

File: localMatchApp/sources/Login.cpp [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  for ( auto &i : i_password ) {
+    if ( isdigit ( i ) ) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool Login::hasCorrectFormat ( const string i_password ) {
+  if ( hasANumber ( i_password ) && hasACapital ( i_password ) && hasASmallLetter ( i_password ) ) {
```
Comment (luvalverde) [inline localMatchApp/sources/Login.cpp]: I wouldn't create separate functions for this.
I would prefer to check the correct password format in this same function, so we only need one FOR loop that checks for digit, capital, lower letter.

File: localMatchApp/CMakeLists.txt [hunk 10]
```diff
@@ -48,4 +48,58 @@ ADD_EXECUTABLE(local_match ${code_SOURCES}
...
-TARGET_LINK_LIBRARIES(local_match Qt5::Widgets)
+TARGET_LINK_LIBRARIES(${PROJECT_NAME} Qt5::Widgets)
+
+###################################################################
+# Gtest section
+
+# Create a variable with the name of the project and "_test"
+SET(TEST_LOCAL_MATCH ${PROJECT_NAME}_test)
+
+# Define the code language as C++ 11
```
Comment (wfallascr) [inline localMatchApp/CMakeLists.txt]: Change this to 14

File: localMatchApp/test/tst_logintest.h [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+#include <iostream>
+#include <string>
+#include "Login.h"
+using namespace testing;
+
+TEST ( LoginTest, TestPassword ) {
+  Login prueba;
+  std::string pass = "Robert0";
+  EXPECT_TRUE ( prueba.hasCorrectFormat ( pass ) );
+  EXPECT_TRUE ( 1 + 1 == 2 );
```
Comment (luvalverde) [inline localMatchApp/test/tst_logintest.h]: why is this here?
for debugging purposes I guess, and you forgot to delete it.

File: localMatchApp/headers/Login.h [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+#include <string>
+class Login {
+ public:
+  bool hasACapital ( const std::string i_password );
+  bool hasASmallLetter ( const std::string i_password );
+  bool hasANumber ( const std::string i_password );
+  bool hasCorrectFormat ( const std::string i_password );
+
+  enum class connectionMessage {
```
Comment (wfallascr) [inline localMatchApp/headers/Login.h]: Class should start with a capital letter

File: localMatchApp/sources/Login.cpp [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+  // Disconnect to database.
+}
+
+Login::connectionMessage Login::signUp ( const string i_user, const string i_password,
+                                         const string i_repeatedPassword ) {
+  // Send information to database.
+  string userInDatabase = "";
+
+  if ( hasCorrectFormat ( i_password ) && hasCorrectFormat ( i_repeatedPassword ) ) {
+    if ( i_user.compare ( userInDatabase ) == 0 ) {
```
Comment (wfallascr) [inline localMatchApp/sources/Login.cpp]: Refactor this, too many nested if/else statements

File: localMatchApp/test/tst_logintest.h [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+
+#include <gmock/gmock-matchers.h>
+#include <gtest/gtest.h>
+#include <iostream>
+#include <string>
+#include "Login.h"
+using namespace testing;
+
+TEST ( LoginTest, TestPassword ) {
+  Login prueba;
```
Comment (luvalverde) [inline localMatchApp/test/tst_logintest.h]: Why does the test logic is inside a .h file?

File: localMatchApp/sources/Login.cpp [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+Login::connectionMessage Login::entry ( const string i_user, const string i_password ) {
+  // Search the user and compare password
+  // Query asking for password based on user
+  string userInDatabase = "";
+  string passInDatabase = "";
+
+  if ( i_user != userInDatabase ) {
+    return connectionMessage::doesntExists;
```
Comment (wfallascr) [inline localMatchApp/sources/Login.cpp]: Try having fewer returns in this function. The ideal case is to have just one return at the end of the function. You could, for example, change the value of a variable in the if statement and return that variable at the end.

File: localMatchApp/sources/Login.cpp [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+  } else if ( i_password != passInDatabase ) {
+    return connectionMessage::wrongPassword;
+  } else {
+    return connectionMessage::goodEntry;
+  }
+}
+
+bool Login::hasACapital ( const string i_password ) {
+  for ( auto &i : i_password ) {
```
Comment (wfallascr) [inline localMatchApp/sources/Login.cpp]: Add const to variable i

File: localMatchApp/test/tst_logintest.h [hunk 17]
```diff
@@ -0,0 +1,15 @@
+#pragma once
+
+#include <gmock/gmock-matchers.h>
+#include <gtest/gtest.h>
+#include <iostream>
+#include <string>
+#include "Login.h"
+using namespace testing;
+
+TEST ( LoginTest, TestPassword ) {
```
Comment (luvalverde) [inline localMatchApp/test/tst_logintest.h]: Provide a description for this test.
Also check negative cases like a password not containing a digit, lower or capital letter.

File: localMatchApp/sources/Login.cpp [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+    if ( isupper ( i ) ) {
+      return true;
```
Comment (wfallascr) [inline localMatchApp/sources/Login.cpp]: Same as before, remove multiple returns when possible. Check the rest of your code

**Review by luvalverde — CHANGES_REQUESTED** [review]
I added some comments @Robchaves

File: localMatchApp/test/tst_logintest.h (hunk 11 above)
Comment (wfallascr) [inline localMatchApp/test/tst_logintest.h]: Clean up this

**Review by wfallascr — CHANGES_REQUESTED** [review]

File: localMatchApp/CMakeLists.txt (hunk 1 above)
Comment (Robchaves) [reply localMatchApp/CMakeLists.txt]: Done.

File: localMatchApp/test/tst_logintest.h (hunk 14 above)
Comment (Robchaves) [reply localMatchApp/test/tst_logintest.h]: Because there is a .cpp file that includes the test and runs it, so everyone can create a test file .h and then add it in the .cpp file.

File: localMatchApp/headers/Login.h (hunk 12 above)
Comment (Robchaves) [reply localMatchApp/headers/Login.h]: Done.

File: localMatchApp/CMakeLists.txt (hunk 10 above)
Comment (Robchaves) [reply localMatchApp/CMakeLists.txt]: Done.

File: localMatchApp/CMakeLists.txt [hunk 19]
```diff
@@ -48,4 +48,58 @@ ADD_EXECUTABLE(local_match ${code_SOURCES}
...
+	${GTestSrc}/src/gtest-all.cc
+	${GMockSrc}/src/gmock-all.cc
+	${test_SOURCES}
+	${code_HEADERS_MOC}
+	)
+
+ADD_TEST(${TEST_LOCAL_MATCH} COMMAND ${TEST_LOCAL_MATCH})
+TARGET_LINK_LIBRARIES(${TEST_LOCAL_MATCH} PRIVATE Threads::Threads Qt5::Widgets)
+###################################################################
+
```
Comment (luvalverde) [inline localMatchApp/CMakeLists.txt]: remove extra blank lines

File: localMatchApp/sources/Login.cpp (hunk 6 above)
Comment (Robchaves) [reply localMatchApp/sources/Login.cpp]: Done.

File: localMatchApp/sources/Login.cpp (hunk 5 above)
Comment (Robchaves) [reply localMatchApp/sources/Login.cpp]: Done.

File: localMatchApp/sources/Login.cpp (hunk 4 above)
Comment (Robchaves) [reply localMatchApp/sources/Login.cpp]: Done.

File: localMatchApp/sources/Login.cpp (hunk 3 above)
Comment (Robchaves) [reply localMatchApp/sources/Login.cpp]: Done.

File: localMatchApp/headers/Login.h (hunk 2 above)
Comment (Robchaves) [reply localMatchApp/headers/Login.h]: Done.

File: localMatchApp/sources/Login.cpp (hunk 9 above)
Comment (Robchaves) [reply localMatchApp/sources/Login.cpp]: Done.

File: localMatchApp/test/tst_logintest.h (hunk 11 above)
Comment (Robchaves) [reply localMatchApp/test/tst_logintest.h]: Deleted.

File: localMatchApp/test/tst_logintest.h (hunk 17 above)
Comment (Robchaves) [reply localMatchApp/test/tst_logintest.h]: Done.

File: localMatchApp/sources/Login.cpp (hunk 16 above)
Comment (Robchaves) [reply localMatchApp/sources/Login.cpp]: Done.

File: localMatchApp/sources/Login.cpp (hunk 13 above)
Comment (Robchaves) [reply localMatchApp/sources/Login.cpp]: Done.

**Comment by Robchaves** [comment]
I've deleted the login class to include it later in another PR.
PR back to you.
@luvalverde  @wfallascr

**Review by luvalverde — APPROVED** [review]
Looks good for me :)

File: localMatchApp/test/tst_logintest.h (hunk 11 above)
Comment (Robchaves) [reply localMatchApp/test/tst_logintest.h]: Done.

File: localMatchApp/CMakeLists.txt (hunk 19 above)
Comment (Robchaves) [reply localMatchApp/CMakeLists.txt]: Done.

File: localMatchApp/CMakeLists.txt [hunk 20]
```diff
@@ -1,8 +1,10 @@
 CMAKE_MINIMUM_REQUIRED(VERSION 3.1)
 
 # Project name and language of it
-PROJECT(local_match
-  LANGUAGES CXX)
+PROJECT(local_match)
+
+# Eneable standard C++ 14 compiler
```
Comment (luvalverde) [inline localMatchApp/CMakeLists.txt]: Enable

**Review by luvalverde — CHANGES_REQUESTED** [review]

File: localMatchApp/CMakeLists.txt (hunk 20 above)
Comment (Robchaves) [reply localMatchApp/CMakeLists.txt]: Done.

**Review by luvalverde — APPROVED** [review]

**Review by wfallascr — APPROVED** [review]
